File: api/internal/db/locations.go
package db

import (
	"errors"

	"github.com/google/uuid"
)

var (
	ErrDupLocationName = errors.New("location name already exists")
	ErrNoLocation      = errors.New("location not found")
)

type AddLocationReq struct {
	Title            string
	CountryCode      string
	PostalAddress    string
	PostalCode       string
	OpenStreetMapURL string
	CityAka          []string

	EmployerID uuid.UUID

	// Currently unused, but will be used in the future for audit logs
	OrgUserID uuid.UUID
}

type DefunctLocationReq struct {
	Title string

	EmployerID uuid.UUID

	// Currently unused, but will be used in the future for audit logs
	OrgUserID uuid.UUID
}

type GetLocByNameReq struct {
	Title string

	EmployerID uuid.UUID
}

type GetLocationsReq struct {
	States []string

	PaginationKey string
	Limit         int

	EmployerID uuid.UUID
}

type RenameLocationReq struct {
	OldTitle string
	NewTitle string

	EmployerID uuid.UUID

	// Currently unused, but will be used in the future for audit logs
	OrgUserID uuid.UUID
}

type UpdateLocationReq struct {
	Title            string
	CountryCode      string
	PostalAddress    string
	PostalCode       string
	OpenStreetMapURL string
	CityAka          []string

	EmployerID uuid.UUID

	// Currently unused, but will be used in the future for audit logs
	OrgUserID uuid.UUID
}

File: typespec/employer/locations.ts
import { City, CountryCode } from '../common/common';

export type LocationState = 'ACTIVE_LOCATION' | 'DEFUNCT_LOCATION';

export const LocationStates = {
    ACTIVE: 'ACTIVE_LOCATION' as LocationState,
    DEFUNCT: 'DEFUNCT_LOCATION' as LocationState,
} as const;

export interface Location {
    title: string;
    country_code: CountryCode;
    postal_address: string;
    postal_code: string;
    openstreetmap_url?: string;
    city_aka?: City[];
    state: LocationState;
}

export interface AddLocationRequest {
    title: string;
    country_code: CountryCode;
    postal_address: string;
    postal_code: string;
    openstreetmap_url?: string;
    city_aka?: City[];
}

export interface DefunctLocationRequest {
    title: string;
}

export interface GetLocationRequest {
    title: string;
}

export interface GetLocationsRequest {
    states?: LocationState[];
    pagination_key?: string;
    limit?: number;
}

export interface RenameLocationRequest {
    old_title: string;
    new_title: string;
}

export interface UpdateLocationRequest {
    title: string;
    country_code: CountryCode;
    postal_address: string;
    postal_code: string;
    openstreetmap_url?: string;
    city_aka?: City[];
}

export function isValidLocationState(state: string): state is LocationState {
    return Object.values(LocationStates).includes(state as LocationState);
} 